File: python/aethernet/uri/router.py
# SPDX-License-Identifier: MIT

"""Dispatches an incoming ``aether://`` URI to the registered handler for its route.

The router is per-app — each app constructs one with its own
:class:`HandlerManifest`.

Lifecycle
---------
1. App startup: build a :class:`HandlerManifest` describing every route the app
   accepts.
2. App startup: register an async callback per :class:`HandlerDescriptor` via
   :meth:`AetherURIRouter.register`.
3. At runtime: when a URI is received (incoming intent, deep link, or in-mesh
   dispatch), call :meth:`AetherURIRouter.dispatch` to invoke the right callback.

Example
-------
::

    manifest = HandlerManifest("aether.media", handlers=(
        HandlerDescriptor("profile", description="Get profile."),
        HandlerDescriptor("content", "{hash}", description="Fetch content."),
    ))
    router = AetherURIRouter(manifest)

    async def on_content(ctx: DispatchContext) -> None:
        await content_service.request(ctx.route_parameters["hash"])

    router.register(manifest.handlers[1], on_content)
    await router.dispatch("aether://KXJB7-MN2P4/content/sha256-abc")
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Awaitable, Callable, Mapping, Union

from aethernet.uri.manifest import HandlerDescriptor, HandlerManifest
from aethernet.uri.uri import AetherURI, parse


# A handler callback receives one DispatchContext and returns an awaitable.
HandlerCallback = Callable[["DispatchContext"], Awaitable[None]]


@dataclass(frozen=True)
class DispatchContext:
    """Context delivered to a registered URI handler when its route matches.

    Attributes
    ----------
    uri:
        The original URI being dispatched.
    handler:
        The matched descriptor.
    route_parameters:
        Route parameters captured from the path template (e.g. ``{"hash": "abc"}``).
    """

    uri: AetherURI
    handler: HandlerDescriptor
    route_parameters: Mapping[str, str]


class AetherURIRouter:
    """Reference in-process router.

    Async- and thread-safe registration via an :class:`asyncio.Lock`.
    """

    def __init__(self, manifest: HandlerManifest) -> None:
        if manifest is None:
            raise ValueError("Manifest is null.")
        self._manifest = manifest
        self._handlers: dict[HandlerDescriptor, HandlerCallback] = {}
        self._lock = asyncio.Lock()

    @property
    def manifest(self) -> HandlerManifest:
        return self._manifest

    # ------------------------------------------------------------------
    # Registration
    # ------------------------------------------------------------------

    def register(self, descriptor: HandlerDescriptor, callback: HandlerCallback) -> None:
        """Register *callback* for *descriptor*.

        The descriptor must be one present in :attr:`manifest`. Re-registering
        replaces the existing callback.

        Note
        ----
        Registration is synchronous so it can be called from app startup code
        outside any event loop. Dispatch itself takes the lock.
        """
        if descriptor is None:
            raise ValueError("Descriptor is null.")
        if callback is None:
            raise ValueError("Handler is null.")
        if descriptor not in self._manifest.handlers:
            raise ValueError(f"Descriptor '{descriptor.name}' is not in the manifest.")
        # No lock needed for the registration write — Python dict assignment is
        # atomic, and the dispatch path takes the lock when *reading*.
        self._handlers[descriptor] = callback

    # ------------------------------------------------------------------
    # Dispatch
    # ------------------------------------------------------------------

    async def dispatch(self, uri: Union[AetherURI, str]) -> bool:
        """Resolve and dispatch *uri*.

        Returns ``True`` iff a registered callback was invoked. If no handler
        is registered (or no manifest entry matches) returns ``False``.
        Handler exceptions propagate to the caller.

        Parameters
        ----------
        uri:
            Either a parsed :class:`AetherURI` or a string. A string is parsed
            via :func:`aethernet.uri.parse` and an :class:`AetherURIError` is
            raised on parse failure.
        """
        parsed: AetherURI = parse(uri) if isinstance(uri, str) else uri

        resolved = self._manifest.resolve(parsed)
        if resolved is None:
            return False
        descriptor, captures = resolved

        async with self._lock:
            callback = self._handlers.get(descriptor)
        if callback is None:
            return False

        ctx = DispatchContext(uri=parsed, handler=descriptor, route_parameters=captures)
        await callback(ctx)
        return True

File: swift/Sources/AetherNetProtocol/URI/Router.swift
// SPDX-License-Identifier: MIT

/// Context delivered to a registered URI handler when its route matches.
/// Carries the original URI plus any captured route parameters.
public struct DispatchContext: Sendable {

    public let uri: AetherUri
    public let handler: HandlerDescriptor
    public let routeParameters: [String: String]

    public init(
        uri: AetherUri,
        handler: HandlerDescriptor,
        routeParameters: [String: String]
    ) {
        self.uri = uri
        self.handler = handler
        self.routeParameters = routeParameters
    }
}

/// Async callback fired when a URI matches a registered handler.
public typealias AetherUriHandler = @Sendable (DispatchContext) async throws -> Void

/// Dispatches an incoming `aether://` URI to the registered handler for its
/// route. The router is per-app — each app constructs one with its own manifest.
///
/// Lifecycle:
///   1. App startup: build a ``HandlerManifest`` describing every route the app accepts.
///   2. App startup: register a callback per ``HandlerDescriptor`` via ``registerHandler(_:handler:)``.
///   3. At runtime: when a URI is received (incoming intent, deep link, or in-mesh dispatch),
///      call ``dispatch(_:)`` or ``dispatch(string:)`` to invoke the right callback.
public protocol AetherUriRouting: Sendable {

    /// The manifest the router resolves against. Nonisolated; safe to read
    /// from any context — actor adopters expose it as `nonisolated let`.
    nonisolated var manifest: HandlerManifest { get }

    /// Registers a callback for a handler descriptor. The descriptor must be one
    /// present in ``manifest``. Re-registering replaces the existing callback.
    func registerHandler(_ descriptor: HandlerDescriptor, handler: @escaping AetherUriHandler) async throws

    /// Resolves and dispatches a URI. Returns `true` if a handler was matched AND
    /// invoked; returns `false` if no handler matched. The handler's errors propagate.
    @discardableResult
    func dispatch(_ uri: AetherUri) async throws -> Bool

    /// Resolves and dispatches a URI given as a string. Throws ``AetherUri/AetherUriError``
    /// if the string fails to parse.
    @discardableResult
    func dispatch(string uri: String) async throws -> Bool
}

/// Reference in-process implementation of ``AetherUriRouting``. Thread-safe via Swift actor.
public final actor AetherUriRouter: AetherUriRouting {

    // MARK: - State

    public nonisolated let manifest: HandlerManifest

    private var handlers: [HandlerDescriptor: AetherUriHandler] = [:]

    // MARK: - Init

    public init(manifest: HandlerManifest) {
        self.manifest = manifest
    }

    // MARK: - Registration

    public func registerHandler(
        _ descriptor: HandlerDescriptor,
        handler: @escaping AetherUriHandler
    ) throws {
        if !manifest.handlers.contains(descriptor) {
            throw AetherUri.AetherUriError.message(
                "Descriptor '\(descriptor.name)' is not in the manifest."
            )
        }
        handlers[descriptor] = handler
    }

    // MARK: - Dispatch

    @discardableResult
    public func dispatch(_ uri: AetherUri) async throws -> Bool {
        guard let resolved = manifest.resolve(uri) else { return false }
        guard let callback = handlers[resolved.0] else { return false }
        let ctx = DispatchContext(
            uri: uri,
            handler: resolved.0,
            routeParameters: resolved.1
        )
        try await callback(ctx)
        return true
    }

    @discardableResult
    public func dispatch(string uri: String) async throws -> Bool {
        let parsed = try AetherUri.parse(uri)
        return try await dispatch(parsed)
    }
}
